rubbish/src/Mode.hs

```diff
@@ -0,0 +1,217 @@
+import           Pos.Util.TimeWarp        (CanJsonLog (..))
+import           Pos.Util.UserSecret      (HasUserSecret (..))
+import           Pos.Util.Util            (HasLens (..), lensOf', postfixLFields)
+import           Pos.WorkMode             (RealMode, RealModeContext (..))
+
+import           Pos.Rubbish.Hacks        (makePubKeyAddressRubbish)
+
+-- | Command execution context.
+data CmdCtx = CmdCtx
+    { ccPeers :: ![NodeId]
+    }
+
+type RubbishSscType = SscGodTossing
+
+type RubbishMode = ReaderT RubbishContext Production
+
+data RubbishContext = RubbishContext
+    { rcRealModeContext :: !(RealModeContext RubbishSscType)
+    , rcCmdCtx          :: !CmdCtx
+    }
+
+makeLensesWith postfixLFields ''RubbishContext
+
+----------------------------------------------------------------------------
+-- Helpers
+----------------------------------------------------------------------------
+
+-- | Get 'CmdCtx' in 'RubbishMode'.
+getCmdCtx :: RubbishMode CmdCtx
+getCmdCtx = view rcCmdCtx_L
+
+-- | Turn 'RealMode' action into 'RubbishMode' action.
+realModeToRubbish :: RealMode RubbishSscType a -> RubbishMode a
+realModeToRubbish action = do
+    rmc <- view rcRealModeContext_L
+    lift $ runReaderT action rmc
+
+----------------------------------------------------------------------------
+-- Boilerplate instances
+----------------------------------------------------------------------------
+
+instance HasSscContext RubbishSscType RubbishContext where
+    sscContext = rcRealModeContext_L . sscContext
+
+instance HasPrimaryKey RubbishContext where
+    primaryKey = rcRealModeContext_L . primaryKey
+
+instance HasReportingContext RubbishContext  where
+    reportingContext = rcRealModeContext_L . reportingContext
+
+instance HasUserSecret RubbishContext where
+    userSecret = rcRealModeContext_L . userSecret
+
+instance HasShutdownContext RubbishContext where
+    shutdownContext = rcRealModeContext_L . shutdownContext
+
+instance HasNodeContext RubbishSscType RubbishContext where
+    nodeContext = rcRealModeContext_L . nodeContext
+
+instance HasSlottingVar RubbishContext where
+    slottingTimestamp = rcRealModeContext_L . slottingTimestamp
+    slottingVar = rcRealModeContext_L . slottingVar
+
+instance {-# OVERLAPPABLE #-}
+    HasLens tag (RealModeContext RubbishSscType) r =>
+    HasLens tag RubbishContext r
+  where
+    lensOf = rcRealModeContext_L . lensOf @tag
+
+instance HasLoggerName' RubbishContext where
+    loggerName = rcRealModeContext_L . loggerName
+
+instance HasSlogContext RubbishContext where
+    slogContext = rcRealModeContext_L . slogContext
+
+instance HasSlogGState RubbishContext where
+    slogGState = rcRealModeContext_L . slogGState
+
+instance HasJsonLogConfig RubbishContext where
+    jsonLogConfig = rcRealModeContext_L . jsonLogConfig
+
+instance (HasCoreConstants, MonadSlotsData ctx RubbishMode)
+      => MonadSlots ctx RubbishMode
+  where
+    getCurrentSlot = getCurrentSlotSum
```

Comment:
To be honest, I don't like what's going in this module. It'll be hard to guarantee consistency with `RealMode` this way, but I plan for `cardano-node-cli` (aka rubbish) to be a drop-in replacement for a node, used in testing and benchmarks. What I propose to do instead is to add a natural transformation:

```
inRealMode  :: RealMode SscGodTossing a -> RubbishMode a
inRealMode = withReaderT rcRealModeContext
```

and then define methods like this:

```
getCurrentSlotBlocking = inRealMode getCurrentSlotBlocking
```

It's possible to add more logic like this:

```
getCurrentSlotBlocking = do
    additionalWork
    inRealMode getCurrentSlotBlocking
```

but you'll know that the behavior of the RealMode is preserved.